### DISCUSSION TITLE
Endpoint to stream CSV

### DISCUSSION THREAD

**Discussion by mtt-artis** [sqlpage/SQLPage#260] [Ideas]
Hi 👋

I was looking for a quick way to create dashboards and SQLpage looks promising!

My users also need to download big CSV results from a postgre db.

The idea is to have an endpoint (like the JSON one) that streams back a CSV file to the client.

I saw the CSV component but I would prefer to trigger the `select` query when users click on the download link and not when the html page is generated.

**Comment by lovasoa** [comment]
Hello!

That would be a nice and useful feature! 
We could even reuse the existing 'csv' component, with the same attributes, and make it generate a raw csv instead of a csv download button when it is used as a header component (before any other component).

We already have a dependency on a csv library and the mechanisms in place to create new header components. 

Do you want to take a stab at implementing it? I'd always be reachable if you need help to navigate around the codebase. The first thing to look at is probably [render.rs](https://github.com/lovasoa/SQLpage/blob/main/src%2Frender.rs#L53-L59).

**Reply by mtt-artis** [reply]
Hello @lovasoa 
I'm considering attempting it, but I'm currently short on time. Do you think it's fairly straightforward?

**Reply by lovasoa** [reply]
I think it's a few hours if you need to get familiar with the codebase first. But even if you don't push it through the finish line, please open a "draft" pull request, I'll be happy to help!

**Reply by mtt-artis** [reply]
I'm not free next week but try the week after

**Comment by lovasoa** [comment]
I just implemented the usage of csv as a header component, the following should work nicely on the main branch:

```sql
select
    'csv' as component, true as bom, ';' as delimiter, '"' as quote, '"' as escape, 'test.csv' as filename;

with recursive rec as (
    select 0 as column1
    union all
    select column1 + 1
    from rec
)
select
    column1 as x,
    column1 * 99999999 as y
from rec
```

This is an infinite query (it never returns), but it works, and efficiently encodes and streams the results in csv to the user's computer.
